### ISSUE TITLE
Graphics problem with tsne algorithm

### ISSUE THREAD

**Issue by JKL-STAR** [jolibrain/deepdetect#1555]
I am implementing the tsne algorithm, my graphs are not identical with those in the package. Could anyone help me resolve this issue.
![tsne](https://github.com/jolibrain/deepdetect/assets/62751448/43577757-0d37-4407-897f-c99a95515896)

grid_search <- function(diff_i, i, perplexity) {
  # Helper function to obtain σ's based on user-specified perplexity.

  result <- Inf  # Set the first result to be infinity
  σ=0
  norm <- apply(diff_i, 1, function(x) sqrt(sum(x^2)))  # Calculate norms
  std_norm <- sd(norm)  # Use standard deviation of norms to define the search space

  for (σ_search in seq(0.01 * std_norm, 5 * std_norm, length.out = 200)) {
    # Equation 1 Numerator
    p <- exp(-norm^2 / (2 * σ_search^2))
    
    # Set p = 0 when i = j
    p[i] <- 0
    
    # Equation 1 (ε -> 0)
    ε <- .Machine$double.eps  # Value closest to 0
    p_new <- p / sum(p)
    p_new <- pmax(p_new, ε)
    
    # Shannon Entropy
    H <- -sum(p_new * log2(p_new))
    
    # Get log(perplexity equation) as close to equality
  if (isTRUE(abs(log(perplexity) - H * log(2)) <= result)==TRUE) {
  result <- log(perplexity) - H * log(2)
  σ <- σ_search
  }
  }
  
  return(σ)
}

P_ij <- function(X, sigmas) {
  n <- nrow(X)
  pij <- matrix(0, n, n)
  
  for (i in 1:n) {
    diff <- matrix(rep(X[i, ], times = nrow(X)), ncol = ncol(X), byrow = TRUE) - X
    pij[i, ] <- exp(-apply(diff^2, 1, sum) / (2 * sigmas[i]^2))
    diag(pij) <- 0
    pij[i, ] <- pij[i, ] / sum(pij[i, ])
  }
  
  pij <- (pij + t(pij)) / (2 * n)
  ε <- .Machine$double.eps
  pij <- pmax(pij, ε)
  
  return(pij)
}

gradient <- function(P, Q, y) {
  n <- nrow(P)
  m <- ncol(y)
  gradient <- matrix(0, n, m)
  C <- P - Q
  
  for (i in 1:n) {
    diff <- matrix(rep(y[i, ], times = nrow(y)), ncol = ncol(y), byrow = TRUE) - y
    gradient[i, ] <- 4 * C[i, ] %*% (diff * ((1 + apply(diff^2, 1, sum))^(-1)))
  }
  
  return(gradient)
}

Qij <- function(y) {
  n <- nrow(y)
  Qij <- matrix(0, n, n)
  
  for (i in 1:n) {
    diff <- matrix(rep(y[i, ], times = nrow(y)), ncol = ncol(y), byrow = TRUE) - y
    Qij[i, ] <- (1 + apply(diff^2, 1, sum))^(-1)
  }
  
  diag(Qij) <- 0
  Qij <- Qij / sum(Qij)
  ε <- .Machine$double.eps
  Qij <- pmax(Qij, ε)
  
  return(Qij)
}

tsne3 <- function(X, perplexity, dim, n_iter, learning_rate, random_state, early_exaggeration) {
  n <- nrow(X)
  
  sigmas <- numeric(n)
  for (i in 1:n) {
    diff <- matrix(rep(X[i, ], times = nrow(X)), ncol = ncol(X), byrow = TRUE) - X
    sigmas[i] <- grid_search(diff, i, perplexity)
  }
  
  Pij <- P_ij(X, sigmas)
  
  m <- function(t) {
    if (t < 250) {
      return(0.5)
    } else {
      return(0.8)
    }
  }
  
  set.seed(random_state)
  y <- matrix(rnorm(n * dim, mean = 0, sd = 1e-4), nrow = n, ncol = dim)
  Y <- list(matrix(0, nrow = n, ncol = dim), y)
  
  for (t in 1:n_iter) {
    if (t < 250) {
      early_exaggeration <- early_exaggeration
    } else {
      early_exaggeration <- 1
    }
    
    Q <- Qij(Y[[length(Y)]])
    y <- Y[[length(Y)]] - learning_rate * gradient(early_exaggeration * Pij, Q, Y[[length(Y)]]) + m(t) * (Y[[length(Y)]] - Y[[length(Y) - 1]])
    Y <- c(Y, list(y))
    
    if (t %% 100 == 0 || t == 1) {
      cost <- sum(Pij * log(Pij / Q))
      print(paste("Iteration", t, ": Value of Cost Function is", cost))
      #message("Epoch: Iteration #",iter," error is: ",cost)
    }
  }
  
  return(y)
}

res1= tsne3(X=load_digits, perplexity=45, dim=2, n_iter=500, learning_rate=200, random_state = 1000, early_exaggeration = 4)
data1=data.frame(res1)
data1$target=digits$target
ggplot(data1) +
  aes(x = X1, y = X2, color = as.factor(target)) +
  geom_point() + scale_fill_brewer(palette = "Set1") +
  labs(title = "dataset: load_digits, tsne JAMES")

**Comment by fantes** [comment]
Does not seem to be a deepdetect issue
